src/lkl/setup.c

```diff
@@ -646,14 +633,17 @@ static void lkl_mount_disk(
             disk->key[i] = rand();
     }
 
-    int lkl_trace_lkl_syscall_bak = sgxlkl_trace_lkl_syscall;
-    int lkl_trace_internal_syscall_bak = sgxlkl_trace_internal_syscall;
+    const sgxlkl_trace_config_t* tcfg = &sgxlkl_enclave_state.config->trace;
+
+    const int lkl_trace_lkl_syscall_bak = tcfg->lkl_syscall;
+    const int lkl_trace_internal_syscall_bak = tcfg->internal_syscall;
 
-    if ((sgxlkl_trace_lkl_syscall || sgxlkl_trace_internal_syscall) &&
+    if ((tcfg->lkl_syscall || tcfg->internal_syscall) &&
         (disk->roothash || is_encrypted_cfg(disk)))
     {
-        sgxlkl_trace_lkl_syscall = 0;
-        sgxlkl_trace_internal_syscall = 0;
+        // Check: const-casts acceptable?
+        *((bool*)&tcfg->lkl_syscall) = 0;
```

Comment:
Pointer l-value casts make me really nervous.  Why don't we just propagate the flags from the config structure to the existing globals early on, rather than making changes everywhere to directly use the config structure and then needing to modify const fields?

Reply:
Me too, that's why I put the check here. I want the config to be const, but at this particular point the intention is really to modify the config. I don't want to keep the static globals, all settings should be in the config, i.e. for security reviews I'd like to have all state in `sgxlkl_enclave_state`, including the const config. Is it really necessary that we disable the traces here?

Reply:
As the comment explains, we disable tracing at this point so that we don't flood the logs.  The change is conflating the configuration state of the logging with the current state of the logging.  In the old version, the configuration state of the logging was in immutable environment variables (which was not a great place for it to be), the current state of the logging was in globals.  Now, we only have one place and we are using the configuration (which I believe should be immutable) to represent the current state.

Ideally, we'd allocate the config structure on a page boundary and do an ocall to make it read-only to prevent accidental modification.

Reply:
I can imagine other scenarios where we'd want to dynamically turn logging on and off (and, longer term, we probably want to expose APIs that we can poke in the debugger to do this), so would much rather we kept the globals.  I don't know what you use to commit things, but most git GUIs make it easy to revert parts of a change.

Reply:
I was kidding, it's just a few lines :) I would want to keep the globals in `sgxlkl_enclave_state` anyways.

Reply:
Sadly, I can't find a way to change the access permissions of a page. SGX1 does not support `EMODP*` and I can't write to a pre-allocated read-only page from within the enclave. Also, `EADD`/`EEXTEND` are disabled after `EINIT`. Since the config is public knowledge, we could theoretically save it outside of the enclave memory and `mprotect` it, but that doesn't feel like the right thing to do either.

Reply:
We have an ocall for this.

Reply:
Yes, there's the mprotect ocall, but that won't work on enclave pages, only on host pages. Are you suggesting that we should save it in host memory?

Reply:
(It is advisory, the host can ignore it, but we're using it for debugging and defence in depth here)

Reply:
My experiments tell me that at least some part of this does indeed work. However, Intel says:

> mprotect will modify the OS permissions on that page , but not the Enclave Page Cache Map (EPCM) permissions. You need EMOD instructions to modify the EPCM permissions on that page.

(https://community.intel.com/t5/Intel-Software-Guard-Extensions/Can-I-dynamically-change-an-SGX-memory-page-s-permission-bits/td-p/1130989)

Reply:
Oh I see what you mean, yeah, if it's allowed to skip it, but happens not to, it's still useful for debugging.

Reply:
Done and it seems to work!